File: combo4-agentic-engineering-in-practice/legacy-service-csharp/src/LegacyService/LoggingSetup.cs
using System.Globalization;

namespace LegacyService;

// LoggingSetup.cs -- shared logging config for OrderBase.
//
// Log lines go to stdout AND logs/app-YYYY-MM-DD.log (when a logs/ dir
// exists in the working directory -- prod boxes have one, CI doesn't).
// The format is FROZEN: the metrics pusher (see DOCS/INSTRUCTIONS.md)
// greps these lines every minute. Change it and the dashboards go dark.
//
// Hand-rolled on purpose -- no ILogger<T>, no DI. `GetLogger(name)` mirrors
// the stdlib `logging.getLogger(name)` call in the Python original.

public static class LoggingSetup
{
    private const string LogDir = "logs";
    private static bool _configured;
    private static StreamWriter? _fileWriter;

    public static void Setup()
    {
        if (_configured)
        {
            return;
        }
        if (Directory.Exists(LogDir))
        {
            var fname = Path.Combine(LogDir, $"app-{DateTime.Now.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)}.log");
            _fileWriter = new StreamWriter(fname, append: true) { AutoFlush = true };
        }
        _configured = true;
    }

    public static Logger GetLogger(string name) => new(name);

    internal static void Emit(string level, string name, string line)
    {
        var ts = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss,fff", CultureInfo.InvariantCulture);
        var full = $"{ts} {level} {name} {line}";
        Console.WriteLine(full);
        _fileWriter?.WriteLine(full);
    }
}

public class Logger
{
    private readonly string _name;

    internal Logger(string name)
    {
        _name = name;
    }

    public void Info(string message, params object?[] args)
        => LoggingSetup.Emit("INFO", _name, Format(message, args));

    public void Warning(string message, params object?[] args)
        => LoggingSetup.Emit("WARNING", _name, Format(message, args));

    private static string Format(string message, object?[] args)
        => args.Length > 0 ? string.Format(CultureInfo.InvariantCulture, message, args) : message;
}

File: combo2-advanced-agentic-engineering/legacy-service-java/src/main/java/ai/octoco/legacyservice/LoggingSetup.java
package ai.octoco.legacyservice;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

// LoggingSetup.java -- shared logging config for OrderBase.
//
// Log lines go to stdout AND logs/app-YYYY-MM-DD.log (when a logs/ dir
// exists in the working directory -- prod boxes have one, CI doesn't).
// The format is FROZEN: the metrics pusher (see DOCS/INSTRUCTIONS.md)
// greps these lines every minute. Change it and the dashboards go dark.
//
// Hand-rolled on purpose -- no SLF4J, no Logback, no DI. getLogger(name)
// mirrors the stdlib logging.getLogger(name) call in the Python original.

public final class LoggingSetup {

    private static final String LOG_DIR = "logs";
    private static final DateTimeFormatter FILE_DATE =
            DateTimeFormatter.ofPattern("yyyy-MM-dd", Locale.ROOT);
    private static final DateTimeFormatter LINE_TIMESTAMP =
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss,SSS", Locale.ROOT);

    private static boolean configured;
    private static PrintWriter fileWriter;

    private LoggingSetup() {}

    public static synchronized void setup() {
        if (configured) {
            return;
        }
        Path logDir = Path.of(LOG_DIR);
        if (Files.isDirectory(logDir)) {
            String fname = "app-" + LocalDateTime.now().format(FILE_DATE) + ".log";
            try {
                fileWriter = new PrintWriter(new FileWriter(logDir.resolve(fname).toFile(), true), true);
            } catch (IOException e) {
                // If the log file can't be opened, keep going with stdout only --
                // same as the C# port silently accepting a missing logs/ dir.
                fileWriter = null;
            }
        }
        configured = true;
    }

    public static Logger getLogger(String name) {
        return new Logger(name);
    }

    static synchronized void emit(String level, String name, String line) {
        String ts = LocalDateTime.now().format(LINE_TIMESTAMP);
        String full = ts + " " + level + " " + name + " " + line;
        System.out.println(full);
        if (fileWriter != null) {
            fileWriter.println(full);
        }
    }
}
